backend/app/main.py

```diff
@@ -28,59 +31,102 @@
     ],
 )
 
+oauth2_scheme = HTTPBearer()
 
-@app.get('/listings', response_model=List[ListingsGetResponseItem])
-def get_listings() -> List[ListingsGetResponseItem]:
+### Decode JWT Token (Protect Routes)
+async def get_current_user(token: str = Depends(oauth2_scheme)):
     """
-    Retrieve all listings
+    Decode the JWT token and return the user if valid.
     """
-    pass
+    try:
+        token = token.credentials  # Extract raw token from HTTPBearer object
+        payload = decode(token, JWT_SECRET, algorithms=["HS256"])
+        email: str = payload.get("email")
+        user_id: str = payload.get("id")
 
+        if email is None or user_id is None:
+            raise HTTPException(status_code=401, detail="Invalid token")
 
-@app.post(
-    '/listings', response_model=None, responses={'201': {'model': ListingsPostResponse}}
-)
-def post_listings(body: ListingsPostRequest) -> Optional[ListingsPostResponse]:
+        return {"email": email, "id": user_id}
+
+    except exceptions.ExpiredSignatureError:
+        raise HTTPException(status_code=401, detail="Token expired")
+    except exceptions.DecodeError:
+        raise HTTPException(status_code=401, detail="Invalid token")
+
+### User Authentication (Verify Email & Password)
+async def authenticate_user(username: str, password: str):
+    """
+    Check if the user exists and verify the password.
+    """
+    user = db.users.find_one({"email": username})  # ✅ Async database query
+
+    if not user or not pbkdf2_sha256.verify(password, user["password"]):
+        return None
+    return user
+
+### Define JWT Token Response
+class TokenResponse(BaseModel):
+    access_token: str  # Token string
+    token_type: str  # Always "bearer"
+
+### Login Request Model (Replaces OAuth2PasswordRequestForm)
+class LoginRequest(BaseModel):
+    username: str
+    password: str
+
+### Generate JWT Token on Successful Login
+@app.post('/login', response_model=TokenResponse)
+async def login(body: LoginRequest):
+    """
+    Authenticate user and return a JWT token.
+    """
+    user = await authenticate_user(body.username, body.password)
+
+    if not user:
+        raise HTTPException(status_code=401, detail="Incorrect email or password")
+
+    token = jwt.encode({"email": user["email"], "id": str(user["_id"])}, JWT_SECRET, algorithm="HS256")
+    return TokenResponse(access_token=token, token_type="bearer")
+
+### Public Route: Retrieve All Listings (No Authentication Required)
+@app.get('/listings', response_model=List[ListingsGetResponseItem])
```

Comment:
Let's use OpenAPI to regenerate the boilerplates for this after the authentication OpenAPI spec is written.